nix/eval-machine-info.nix

```diff
@@ -11,104 +11,134 @@
 let
   call = x: if builtins.isFunction x then x args else x;
 
-  # Copied from nixpkgs to avoid <nixpkgs> import
-  optional = cond: elem: if cond then [elem] else [];
-
   zipAttrs = set: builtins.listToAttrs (
     map (name: { inherit name; value = builtins.catAttrs name set; }) (builtins.concatMap builtins.attrNames set));
 
-  flakeExpr = (builtins.getFlake flakeUri).outputs.nixopsConfigurations.default;
-
-  networks =
-    let
-      getNetworkFromExpr = networkExpr:
-        (call (import networkExpr)) // { _file = networkExpr; };
-
-      exprToKey = key: { key = toString key; };
-
-      networkExprClosure = builtins.genericClosure {
-        startSet = map exprToKey networkExprs;
-        operator = { key }: map exprToKey ((getNetworkFromExpr key).require or []);
-      };
-    in
-      map ({ key }: getNetworkFromExpr key) networkExprClosure
-      ++ optional (flakeUri != null)
-        ((call flakeExpr) // { _file = "<${flakeUri}>"; });
-
-  network = zipAttrs networks;
-
-  evalConfig =
-    if flakeUri != null
-    then
-      if network ? nixpkgs
-      then (builtins.head (network.nixpkgs)).lib.nixosSystem
-      else throw "NixOps network must have a 'nixpkgs' attribute"
-    else import (pkgs.path + "/nixos/lib/eval-config.nix");
-
-  pkgs = if flakeUri != null
-    then
-      if network ? nixpkgs
-      then (builtins.head network.nixpkgs).legacyPackages.${system}
-      else throw "NixOps network must have a 'nixpkgs' attribute"
-    else (builtins.head (network.network)).nixpkgs or (import <nixpkgs> { inherit system; });
+  flake = builtins.getFlake flakeUri;
+  flakeExpr = (builtins.tryEval flake.outputs.nixopsConfigurations.default).value or { };
+
+  nixpkgsBoot = <nixpkgs> ; # this will be replaced on install by nixops' nixpkgs input
+  libBoot = import "${nixpkgsBoot}/lib";
+
+  evalBoot = libBoot.evalModules {
+    specialArgs = args;
+    modules = [
+      {
+        options.nixpkgs = libBoot.mkOption {
+          type = libBoot.types.path;
+          description = "Path to the nixpkgs instance used to buld the machines.";
+          defaultText = libBoot.literalDocBook "The 'nixpkgs' input to either the provided flake or nixops' own.";
+          default = (builtins.tryEval flake.inputs.nixpkgs).value or nixpkgsBoot;
+        };
+        config._module.freeformType = with libBoot.types;attrsOf anything;
+      }
+      flakeExpr
+    ] ++ networkExprs;
+  };
 
-  inherit (pkgs) lib;
+  inherit (evalBoot.config) nixpkgs;
+
+  pkgs = nixpkgs.legacyPackages.${system} or import nixpkgs { inherit system; };
+  lib = nixpkgs.lib or pkgs.lib or libBoot;
+  inherit (lib) mkOption types removeAttrs;
+
+in
+rec {
+
+  network = (evalBoot.extendModules {
```

Comment:
This `extendModules` comes from `libBoot`, so this should use `lib.evalModules` from scratch instead.

(Sorry if I'm repetitive; made some commit comments by accident earlier)